File: guests/ocaml/gallery.ml
(* The gallery scene, OCaml port — guests/rust/gallery.rs,
   tools/scenes/gallery.steps. *)

open Kaya_wire
open Kaya_app

(* A 2x2 RGB PNG, 75 bytes, embedded as source. *)
let test_png =
  Bytes.of_string
    "\137\080\078\071\013\010\026\010\000\000\000\013\073\072\068\
     \082\000\000\000\002\000\000\000\002\008\002\000\000\000\253\
     \212\154\115\000\000\000\018\073\068\065\084\120\156\099\248\
     \207\192\192\000\194\012\255\129\000\000\031\238\005\251\011\
     \217\104\139\000\000\000\000\073\069\078\068\174\066\096\130"

let () =
  let app = Kaya_app.create () in

  build app (fun () ->
     let status = signal (Str "urgent: false") in
     let volume = signal (Str "volume: 50%") in
     let pos = signal (F64 0.5) in

     let on_urgent checked =
       write status (Str (Printf.sprintf "urgent: %b" checked))
     in
     let on_volume v =
       (* Integer percent, so every language's formatting agrees. *)
       write volume
         (Str (Printf.sprintf "volume: %d%%"
                 (int_of_float (Float.round (v *. 100.)))))
     in
     (* A programmatic write must NOT come back as an occurrence. *)
     let on_quarter () = write pos (F64 0.25) in

     let root =
       column
         [
           row [ checkbox ~text:"urgent" ~on_toggle:on_urgent; label ~bind:status ];
           row
             [
               slider ~min:0.0 ~max:1.0 ~bind:pos ~on_change:on_volume;
               label ~bind:volume;
               button ~text:"quarter" ~on_click:on_quarter;
             ];
           (* Deliberately invalid bytes: a decode failure reads 0x0. *)
           row
             [
               image ~source:test_png;
               image ~source:(Bytes.of_string "not an image");
             ];
         ]
         ()
     in
     mount root);

  exit (run app)

File: guests/swift/gallery.swift
// The gallery scene, Swift port — guests/rust/gallery.rs,
// tools/scenes/gallery.steps.

import Foundation

/// A 2x2 RGB PNG, 75 bytes, embedded as source.
let testPNG = Data([
    137, 80, 78, 71, 13, 10, 26, 10, 0, 0, 0, 13, 73, 72, 68, 82, 0, 0, 0, 2,
    0, 0, 0, 2, 8, 2, 0, 0, 0, 253, 212, 154, 115, 0, 0, 0, 18, 73, 68, 65,
    84, 120, 156, 99, 248, 207, 192, 192, 0, 194, 12, 255, 129, 0, 0, 31, 238,
    5, 251, 11, 217, 104, 139, 0, 0, 0, 0, 73, 69, 78, 68, 174, 66, 96, 130,
])

let app = KayaApp()

app.build { tx in
    let status = tx.signal(.str("urgent: false"))
    let volume = tx.signal(.str("volume: 50%"))
    let pos = tx.signal(.f64(0.5))

    let root = tx.column {
        tx.row {
            tx.checkbox("urgent") { t, checked in
                t.write(status, .str("urgent: \(checked)"))
            }
            tx.label(bind: status)
        }
        tx.row {
            // Integer percent, so every language's formatting agrees.
            tx.slider(min: 0.0, max: 1.0, bind: pos) { t, value in
                t.write(volume, .str("volume: \(Int((value * 100).rounded()))%"))
            }
            tx.label(bind: volume)
            // A programmatic write must NOT come back as an occurrence.
            tx.button("quarter") { t in
                t.write(pos, .f64(0.25))
            }
        }
        tx.row {
            // Invalid bytes read 0x0: decode failure is the placeholder
            // class, never a crash, on every backend.
            tx.image(testPNG)
            tx.image(Data("not an image".utf8))
        }
        // The labelled row: the control's accessibility name IS the
        // label's text, with no a11yLabel of its own.
        tx.labeled("Level") {
            tx.setA11yId(tx.slider(min: 0.0, max: 1.0, value: 0.5), "level")
        }
    }
    tx.mount(root)
}

app.run()
